src/models/transaction.ts

```diff
@@ -0,0 +1,350 @@
+   *
+   * @return {number} Sum of outputs
+   * @memberof Transaction
+   * @inner
+   */
+  getOutputsSum(): number {
+    let sumOutputs = 0;
+    for (const output of this.outputs) {
+      if (output.isAuthority()) {
+        continue
+      }
+      sumOutputs += output.value;
+    }
+    return sumOutputs;
+  }
+
+  /**
+   * Serialize tx to bytes
+   *
+   * @return {Buffer}
+   * @memberof Transaction
+   * @inner
+   */
+  toBytes(): Buffer {
+    let arr: any = []
+    // Serialize first the funds part
+    //
+    // Tx version
+    arr.push(helpers.intToBytes(this.version, 2))
+    if (this.tokens.length) {
+      // Len tokens
+      arr.push(helpers.intToBytes(this.tokens.length, 1))
+    }
+    // Len inputs
+    arr.push(helpers.intToBytes(this.inputs.length, 1))
+    // Len outputs
+    arr.push(helpers.intToBytes(this.outputs.length, 1))
+
+    for (const token of this.tokens) {
+      arr.push(new encoding.BufferReader(token).buf);
+    }
+
+    for (const inputTx of this.inputs) {
+      arr.push(util.buffer.hexToBuffer(inputTx.hash));
+      arr.push(helpers.intToBytes(inputTx.index, 1));
+      if (inputTx.data) {
+        arr.push(helpers.intToBytes(inputTx.data.length, 2));
+        arr.push(inputTx.data);
+      } else {
+        arr.push(helpers.intToBytes(0, 2));
+      }
+    }
+
+    for (const outputTx of this.outputs) {
+      arr.push(outputTx.valueToBytes());
+      // Token data
+      arr.push(helpers.intToBytes(outputTx.tokenData, 1));
+
+      const outputScript = outputTx.createScript();
+      arr.push(helpers.intToBytes(outputScript.length, 2));
+      arr.push(outputScript);
+    }
+
+    if (this.version === CREATE_TOKEN_TX_VERSION) {
+      // Create token tx need to add extra information
+      arr = [...arr, ...this.serializeTokenInfo()];
+    }
+
+    // Now serialize the graph part
+    //
+    // Weight is a float with 8 bytes
+    arr.push(helpers.floatToBytes(this.weight, 8));
+    // Timestamp
+    if (this.timestamp) {
+      arr.push(helpers.intToBytes(this.timestamp, 4))
+    }
+
+    if (this.parents) {
+      arr.push(helpers.intToBytes(this.parents.length, 1))
+      for (const parent of this.parents) {
+        arr.push(util.buffer.hexToBuffer(parent));
+      }
+    } else {
+      // Len parents (parents will be calculated in the backend)
+      arr.push(helpers.intToBytes(0, 1))
+    }
+
+    // Add nonce in the end
+    arr.push(helpers.intToBytes(this.nonce, 4));
+    return util.buffer.concat(arr);
+  }
+
+  /**
+   * Validate transaction information.
+   * For now, we only verify the maximum number of inputs and outputs.
+   *
+   * @throws {MaximumNumberInputsError} If the tx has more inputs than the maximum allowed
+   * @throws {MaximumNumberOutputsError} If the tx has more outputs than the maximum allowed
+   *
+   * @memberof Transaction
+   * @inner
+   */
+  validate() {
+    if (this.inputs.length > MAX_INPUTS) {
```

Comment:
Why don't we add other validations?

Reply:
Which ones? I've added the ones that were already in the lib. We could add almost all validations from full node, like the sum of inputs/outputs and others but is it useful? It's a bunch of code because there are special cases with authorities.

Reply:
I agree that we can do it later. Maybe we should just create an issue to do that.

The very first next step can be to validate transactions with no authority inputs/outputs. This validation is very simple.

Reply:
Issue created: https://github.com/HathorNetwork/hathor-wallet-lib/issues/160